### ISSUE TITLE
Sampled Softmax for Tensorflow Keras

### ISSUE THREAD

**Issue by user06039** [tensorflow/tensorflow#36115]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.1.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

This is a feature request for sampled softmax loss in Tensorflow-2 Keras. The sampled softmax loss is important when dealing with large number of target classes mainly in sequence to sequence models.

Sampled softmax is available in tensorflow via `tf.nn.sampled_softmax` but I couldn't find any resource that can help me to use it with tensorflow keras.

_

> **Note: If anyone have **implemented it**, **please** share the sample code on how you've achieved it. There are 10's of questions in StackOverflow regarding this without any answer. **

_

**Will this change the current api? How?**
No.
**Who will benefit with this feature?**
This will benefit those who are dealing with large number of target classes to predict. In such cases softmax is not a ideal solution.

**Comment by John-8704** [comment]
@fchollet Any progress on this?

**Comment by ronnygeo** [comment]
I got this to work on tf 2.5.0, after weeks of trying, I had to implement a custom training_step and feed in the input, bias and weights to the final layer. The key was figuring out a way to pass in the inputs (forward activation) as a graph for keras model.

Loss definition
```
class SampledSoftmaxLoss:  
  def __init__(self, num_classes, sample_frac=0.5, num_true_labels=1, loss_name="sampled_softmax_loss"): 
     ...

  def loss(self, y_true, data):   
    input, bias, weights = data      
    labels = y_true    
    labels = tf.expand_dims(y_true, -1)
    logits = tf.matmul(input, weights)    
    logits = tf.nn.bias_add(logits, bias)

    return logits, tf.reduce_mean(tf.nn.sampled_softmax_loss(weights=tf.transpose(weights), biases=bias,            labels=labels, inputs=input, num_true=self.num_true_labels, num_sampled = int(self.num_classes * self.sample_frac),            num_classes = self.num_classes))
```

Then I could add this as my train_loss in my model definition:
`self.train_loss=SampledSoftmaxLoss(self.num_classes).loss`

While defining the model, I made two keras models, sharing all the same layers, except the last:
```
    dense2 = L.Dense(100, "relu")(dense1)   
    output_layer = L.Dense(self.num_classes)    
    self.out_weights = output_layer.weights[0]    
    self.out_bias = output_layer.bias
    out = output_layer(dense2)
    self.e2e_model = tf.keras.Model(inp, out)    
    self.model_onebefore = tf.keras.Model(inp, dense2)
```

With this new onebefore model, I can get the output from dense2 (which is the **input** to output_layer), so finally in the train_step I can call the loss function like this:
```
with tf.GradientTape() as tape:      
  input = self.model_onebefore(x, training=True)  # Forward pass      
  y_pred, loss = self.train_loss(y_true, [input, self.out_bias, self.out_weights])
```

I'm also returning the logit from the loss function so I can compute the metrics but its optional. I noticed that the loss is not automatically returned from the train_step now, so I had to create a loss_tracker as [mentioned here](https://www.tensorflow.org/guide/keras/customizing_what_happens_in_fit#going_lower-level).


Im pretty sure I can further clean this up but happy to finally get it working, until the official keras version is out. 
See the [full gist here](https://gist.github.com/ronnygeo/f73aa4a8f4abf2c950f9d6b755780d67) if more questions. Hope this works for you and let me know if I missed something.

**Comment by innat** [comment]
Have you checked 

https://stackoverflow.com/q/56821654/9215780

https://stackoverflow.com/q/47892380/9215780

**Comment by user06039** [comment]
@innat Actually I tried them for seq2seq model, but they didn't work for me. I'll try them one more, since I don't see this coming anytime soon. But having it natively in Keras is not a bad idea, though.

**Comment by Saduf2019** [comment]
@user06039 
Please refer to below links for reference: 
[link](https://www.examplefiles.net/cs/1104466)
[link](https://pretagteam.com/question/sampled-softmax-in-tensorflow-keras)



Please post this issue on keras-team/keras repo.
To know more refer to:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
